### PR TITLE
Add support for user tokens in AuthenticationSettings

### LINKED ISSUE #1154: Support UserTokens for repository authentication
Currently we support authenticating against a package repository with a jwt token by configuring AuthenticationSettings. We should also support User Tokens which can be scoped, generated, and revoked by normal users.

### PR DESCRIPTION
**Pull request by alnlarsen**
Add support of user tokens

* Obsolete Expiration property
* Add property identifying the token kind
* Update doc comments to reflect that this is not necessarily a jwt
* Make expiration check more explicit

Closes #1154

### CODE REVIEW

File: Engine/Authentication/AuthenticationSettings.cs [hunk 1]
```diff
@@ -96,11 +99,18 @@ void PrepareRequest(HttpRequestMessage request, string domain, CancellationToken
                 token = Tokens.FirstOrDefault(t => t.Domain == request.RequestUri.Host);
             if (token != null)
             {
-                if (token.Expiration < DateTime.Now.AddSeconds(10))
+                // If a token is known to be expired, we should not include it in the header. Doing so will cause
+                // all requests to fail against e.g. the package repository because it completely rejects failing credentials.
+                if (token.Kind == TokenInfo.TokenKind.Jwt && token.Claims.TryGetValue("exp", out var exp) && long.TryParse(exp, out var ts))
```
Comment (rmadsen-ks) [inline Engine/Authentication/AuthenticationSettings.cs:104]: Maybe we should just remove this?

File: Engine/Authentication/TokenInfo.cs [hunk 2]
```diff
@@ -54,13 +74,21 @@ public string Domain
...
+        private Dictionary<string, string> getClaims()
+        {
+            if (Kind == TokenKind.Jwt)
+                return _Claims = GetPayload().RootElement.EnumerateObject().ToDictionary(c => c.Name, c => c.Value.ToString());
+            return new Dictionary<string, string>();
+        }
+
         /// <summary> Expiration date of the <see cref="AccessToken"/>. </summary>
         [XmlIgnore]
+        [Obsolete("Don't make assumptions about the kind or content of tokens.")]
```
Comment (rmadsen-ks) [inline Engine/Authentication/TokenInfo.cs:91]:
```suggestion
        [Obsolete("Expiration time is not supported. Future token types will not contain expiration information. ")]
```

File: Engine/Authentication/TokenInfo.cs [hunk 3]
```diff
@@ -10,10 +10,30 @@
...
     /// <summary> 
-    /// Represents a set of Oauth2/OpenID Connect jwt tokens (access and possibly refresh token) that grants access to a given domain.
+    /// Represents a set of Oauth2/OpenID Connect tokens (access and possibly refresh token) that grants access to a given domain.
     /// </summary>
     public class TokenInfo
     {
+        /// <summary>
+        /// An enumeration of known token kinds
+        /// </summary>
+        public enum TokenKind
```
Comment (rmadsen-ks) [inline Engine/Authentication/TokenInfo.cs:20]: Let's make this private for now.

**Review by rmadsen-ks — CHANGES_REQUESTED** [review]
I had a few commets.

File: Engine/Authentication/AuthenticationSettings.cs (hunk 1 above)
Comment (alnlarsen) [reply Engine/Authentication/AuthenticationSettings.cs:104]: Agreed!

File: Engine/Authentication/AuthenticationSettings.cs [hunk 4]
```diff
@@ -87,6 +87,9 @@ public string BaseAddress
             }
         }
 
+        private static readonly TraceSource Log = OpenTap.Log.CreateSource("authentication");
+        private static readonly HashSet<string> Warned = new HashSet<string>();
```
Comment (rmadsen-ks) [inline Engine/Authentication/AuthenticationSettings.cs:91]:
```suggestion

```

File: Engine/Authentication/TokenInfo.cs [hunk 5]
```diff
@@ -54,13 +74,22 @@ public string Domain
...
+            if (Kind == TokenKind.Jwt)
+                return GetPayload().RootElement.EnumerateObject().ToDictionary(c => c.Name, c => c.Value.ToString());
+            return new Dictionary<string, string>();
+        }
+
         /// <summary> Expiration date of the <see cref="AccessToken"/>. </summary>
         [XmlIgnore]
+        [Obsolete("Expiration time is not supported. Future token types may not contain expiration information. " +
+                  "Consider manually decoding the token if you know the specific format.")]
         public DateTime Expiration => unixEpoch.AddSeconds(long.Parse(Claims["exp"]));
```
Comment (rmadsen-ks) [inline Engine/Authentication/TokenInfo.cs:93]: Should we add `if(Kind !=TokenKind.Jwt) ? DateTime.MaxValue` ?

**Comment by alnlarsen** [comment]
Ah, you are right. Please take another look @rmadsen-ks

**Review by rmadsen-ks — APPROVED** [review]
This looks good!

**Pull request merged**